projects/drivers/util/xml.c

```diff
@@ -0,0 +1,281 @@
+	} else {
+		struct xml_attribute **buff = realloc(node->attributes,
+						      (node->attr_cnt + 1) * sizeof(struct xml_attribute*));
+		if (!buff)
+			return FAILURE;
+		node->attributes = buff;
+	}
+	node->attributes[node->attr_cnt] = attribute;
+	node->attr_cnt++;
+
+	return SUCCESS;
+}
+
+/**
+ * create new xml node
+ * @param **node pointer to the new node
+ * @param *name
+ * @return SUCCESS in case of success or negative value otherwise
+ */
+ssize_t xml_create_node(struct xml_node **node, char *name)
+{
+	*node = calloc(1, sizeof(xml_node));
+	if (!(*node))
+		return FAILURE;
+	(*node)->name = calloc(1, strlen(name) + 1);
+	if (!(*node)->name) {
+		free(*node);
+		return FAILURE;
+	}
+	strcpy((*node)->name, name);
+
+	return SUCCESS;
+}
+
+/**
+ * add child node to a parent node
+ * @param *node_parent
+ * @param *node_child
+ * @return SUCCESS in case of success or negative value otherwise
+ */
+ssize_t xml_add_node(struct xml_node *node_parent, struct xml_node *node_child)
+{
+	if (!node_parent->children) {
+		node_parent->children = calloc(1, sizeof(struct xml_node*));
+		if (!node_parent->children)
+			return FAILURE;
+	} else {
+		struct xml_node **buff = realloc(node_parent->children,
+						 (node_parent->children_cnt + 1) * sizeof(struct xml_node*));
+		if (!buff)
+			return FAILURE;
+		node_parent->children = buff;
+	}
+	node_parent->children[node_parent->children_cnt] = node_child;
+	node_parent->children_cnt++;
+
+	return SUCCESS;
+}
+
+/**
+ * delete attribute
+ * @param *attribute
+ * @return SUCCESS in case of success or negative value otherwise
+ */
+ssize_t xml_delete_attribute(struct xml_attribute **attribute)
+{
+	free((*attribute)->name);
+	free((*attribute)->value);
+	free(*attribute);
+
+	return SUCCESS;
+}
+
+/**
+ * delete xml node
+ * @param *node
+ * @return SUCCESS in case of success or negative value otherwise
+ */
+ssize_t xml_delete_node(struct xml_node **node)
+{
+	uint16_t i;
+	for (i = 0; i < (*node)->attr_cnt; i++) {
+		xml_delete_attribute(&(*node)->attributes[i]);
+	}
+	for (i = 0; i < (*node)->children_cnt; i++) {
+		xml_delete_node(&(*node)->children[i]);
+	}
+	free((*node)->name);
+	free((*node)->attributes);
+	free((*node)->children);
+	free(*node);
+
+	return SUCCESS;
+}
+
+/**
+ * print xml tree into a xml document
+ * @param **document
+ * @param *node pointer to parent node, that contains the xml tree
+ * @return SUCCESS in case of success or negative value otherwise
+ */
+ssize_t xml_create_document(struct xml_document **document,
+			    struct xml_node *node)
+{
+	uint8_t i;
+	ssize_t ret;
+	uint32_t len;
+
+	if (!(*document)) {
+		*document = calloc(1, sizeof(xml_document));
+		if (!(*document))
+			return FAILURE;
+	}
+
+	len = strlen(node->name) + 3;
+	char *buff = realloc((*document)->buff, (*document)->index + len);
+	if (!buff)
+		goto error;
+	(*document)->buff = buff;
+	(*document)->index += sprintf(&(*document)->buff[(*document)->index], "<%s ",
+				      node->name);
```

Comment:
Here, if `node->name` is `NULL` this could be a weird error and would cause the index to go backwards. Can that happen?

Reply:
Checking for NULL values is not done anywhere in the code. Should we check  for NULL values here?

Reply:
since this implementation is pretty complex with arrays of pointers it would be wise to check for null.